acvm/src/lib.rs

```diff
@@ -92,7 +97,26 @@ fn first_missing_assignment(
     })
 }
 
-pub trait Backend: SmartContract + ProofSystemCompiler + PartialWitnessGenerator + Default {}
+pub trait Backend:
+    SmartContract + ProofSystemCompiler + PartialWitnessGenerator + ReferenceString + Default
+{
+}
+
+#[async_trait]
+pub trait ReferenceString {
```

Comment:
Why do we need a new trait for this?
I think all backends should implement this interface. If a backend is not using the crs model, it can easily returns an empty vector.

Reply:
SmartContract makes sense, you do not need it to prove and verify.
Reference string is less clear, I guess you can generate a circuit for a backend, without having the CRS, but the ProofSystemCompiler has the preprocess interface, meaning it requires the ReferenceString.
I suggest to have a Generator trait instead which will deal with the CRS, proving key and verification key.

Reply:
But the SmartContract has a dependency on the CRS, so we should probably just throw out any separation and call it all a single Backend trait.

Reply:
I don't see why you should implement SmartContract if you don't need it, so the separation here is good IMHO.
Generator and ProofSystem have different responsibilities, and Generator does not depend on ProofSystem so I am also fine with such separation.

Reply:
A backend requires SmartContract. You must implement all traits or you aren't a Backend.

Reply:
I think this is bad, but it is another topic, and the trait separation cannot hurt, so I think it is a move in a good direction.

Reply:
In an ideal world, we'd make `SmartContract` optional as I imagine a lot of backends will leave this unimplemented. 
My concern with this is maintaining a consistent interface for nargo as currently we wouldn't be able to compile a backend which didn't implement the full `Backend` trait.

I suggest we park this for the pluggable backends push.

Reply:
#280